clients/pkg/promtail/client/manager.go

```diff
@@ -0,0 +1,49 @@
+package client
+
+import (
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/wal"
+)
+
+// Manager manages remote write client instantiation, and connects the related components to orchestrate the flow of api.Entry
+// from the scrape targets, to the remote write clients themselves.
+//
+// Right now it just supports instantiating the WAL writer side of the future-to-be WAL enabled client. In follow-up
+// work, tracked in https://github.com/grafana/loki/issues/8197, this Manager will be responsible for instantiating all client
+// types: Logger, Multi and WAL.
+type Manager struct {
+	clientChannelHandler api.EntryHandler
+}
+
+// NewManager creates a new Manager
+func NewManager(reg prometheus.Registerer, logger log.Logger, walCfg wal.Config) (*Manager, error) {
+	// TODO: refactor this to instantiate all clients types
+	pWAL, err := wal.New(walCfg, logger, reg)
+	if err != nil {
+		return nil, err
+	}
+	walWriter := wal.NewWriter(pWAL, logger)
```

Comment:
new writer should be created in promtails `main.go`, and passed to all the scrape jobs from there

here is where we would create the watcher for each client in a follow up PR

Reply:
I don't think I agree. The components that bring logs into the system and store them to the WAL should be separated from the components that need to read logs from the WAL and send them over the wire to Loki.

The discovery/scrape targets are responsible for reading log files or grabbing logs data from an external system, and thus should probably be managed by a single overarching struct which also either creates the WAL instance and passes an interface for it to those target routines, or takes the interface as part of it's own construction and can pass it to the targets that way.

The client manager struct is then completely separate, and only takes in some interface for receiving logs (which in reality could be just the existing channel passing we do, or the WAL watcher interface/channel we're working on now, as long as both satisfy the same interface) that it can then duplicate and send to all clients that have been started.

Reply:
After taking a second look I agree with you. It's not part of the client manager concern to route entries emitted by the scraping target to the WAL itself. The WAL itself should act as barrier between the scraping side (scraping targets + writing to the WAL) and the remote write side (reading from the WAL + remote writing).

The only part I'm not that convinced is where to the logic of, for the time being, writing entries produced by scraping target to both the WAL and the clients. I feel it's a smell to create a separate components that does this fanout since it will be removed shorty after.

I took a stab at adding this inside the WAL writer, I think it's not a bad place considering it will live there just until we get [this done](https://github.com/grafana/loki/pull/8302).